vg/draw/canvas.go

```diff
@@ -9,14 +9,33 @@ import (
 	"image/color"
 	"math"
 	"strings"
+	"sync"
 
 	"gonum.org/v1/plot/vg"
-	"gonum.org/v1/plot/vg/vgeps"
-	"gonum.org/v1/plot/vg/vgimg"
-	"gonum.org/v1/plot/vg/vgpdf"
-	"gonum.org/v1/plot/vg/vgsvg"
 )
 
+// formats holds the registered canvas image formats
+var formats struct {
+	sync.Mutex
+	m map[string]func(w, h vg.Length) vg.CanvasWriterTo
+}
+
+// RegisterFormat registers an image format for use by NewformattedCanvas.
+// Name is the name of the format, like "jpeg" or "png".
+// New is the construction function to call for the format.
+func RegisterFormat(name string, fn func(w, h vg.Length) vg.CanvasWriterTo) {
+	formats.Lock()
+	defer formats.Unlock()
+	if formats.m == nil {
+		formats.m = make(map[string]func(w, h vg.Length) vg.CanvasWriterTo)
+	}
+	if fn == nil {
```

Comment:
this isn't documented. I think we should document that.
but I must say this feels to me like `RegisterFormat` wants to do too much.
what's the use case, actually? removing a format from the map won't un-import the `vgfoo` package.

Reply:
Requested by @kortschak, so I like to pass the question for the use case to him.

Reply:
The response is a question, "What is the desired behaviour when someone does `RegisterFormat(name, fn)` and `fn` is accidentally `nil`?" If we're OK panicking with that, then I'm fine with removing this, but replace it with a conditional panic here so that it's seen when it happens. The main motivator for me was easing debugging, but that's not a big deal.

Reply:
I'd be fine with panicking too.
`database/sql.Register` does it as well and documents it:
```
$> go doc database/sql.Register 
func Register(name string, driver driver.Driver)
    Register makes a database driver available by the provided name. If Register
    is called twice with the same name or if driver is nil, it panics.
```

Reply:
OK, I would like to be able to call `RegisterFormat` twice though.

Reply:
because you are worried about multiple different packages registering their favorite `vg` backend in their `init` function, which happens to be `"png"` ?
yeah... I guess that could happen.

Reply:
Since we have no control over the order in which packages are initialized in the eventual program, the outcome of "which package will provide `"png"`" is uncertain (at least for the formats registered by the `init()`s). A `panic()` might be cleaner, might it not? 
Hmm, OTOH, for that same reason the author of `main()` might have no immediate control over what packages are initialized further down in the dependency tree, and allowing double registrations for the same formats would at least allow the program to run.
To give another example compared to `database/sql`: [Package `image`](https://golang.org/src/image/format.go#26) uses a slice in which the the first registration would always win, but which would never panic. If a registered function is `nil`, the function parsing the formats returns an error (for us, that would be `NewFormattedCanvas`).

Anyway, no matter which way we decide to have implemented here, I agree that I need to mention it in the  documentation.

Reply:
the final user/dev of `main` has perhaps no control over the order of package initialization (except for the one in the `package main`, of course) but the order is stable, single-threaded and reproducible.
so given a set of packages that depend on each others, the order can be known and (somewhat) standardized:
- https://golang.org/ref/spec#Package_initialization

_ie:_ it's not C++ :)

Reply:
There have been no further comments, so I am going to
- implement and document panic on `nil`, but
- keep allowing format registrations under the same name.